### DISCUSSION TITLE
Occlum 2021 roadmap

### DISCUSSION THREAD

**Discussion by guzongmin** [occlum/occlum#406] [General]
In 2021 is a great year for Occlum, below is the roadmap of the year:
Q1:

- [x] Enable code coverage data collection
- [x] Enable gVisor test suit
- [x] Enable user code, heap, stack base address randomization 
- [x] Support encrypted Occlum image 

Q2:

- [ ] Official release Occlum 1.0
- [x] Rewrite Occlum process/thread scheduler with async/await mechanism
- [ ] Rewrite Occlum network module with Linux IO_uring APIs
- [ ] Support RA-TLS
- [ ] Add more communication channels: Dingding and more?
- [ ] Add more badges: license, release, issues
- [ ] Enable more popular software in Occlum
    - [x] Redis
    - [x] Flink
    - [x] SOFABoot
    - [x] Vault 
    - [x] PyTorch
    - [x] tensorflow

Q3:

- [ ] Add more Occlum documents
    - [ ] Better README file
    - [ ] Installation guide
    - [ ] Contribution guide
    - [ ] multi-language version README and other documents: 中文 and more?
- [ ] Improve Occlum performance to close the native environment 
- [ ] Enable SGX EDMM 
    - [ ] Improve  Occlum loading performance
    - [ ] Improve mprotect with EDMM API
- [ ] Support Occlum auto-config (User do not need to config Occlum.json anymore) 
    - [x] A new tool could help user copy correct libc and other libraries into the right folder of Occlum image
    - [ ] User friendly config file: yaml?  
- [ ] Enable more popular software in Occlum
    - [ ] nginx
    - [x] spark

Q4:

- [ ] Support two more languages
    - [ ] Javascript/Typescript
    - [ ] WebAssembly
- [ ] Support python multi-process APIs
- [x] Support Java multi-process APIs
- [ ] Improve the test code coverage to 85%
- [ ] Enable more popular software in Occlum
    - [ ] hadoop
    - [x] bash
    - [ ] Kafka
    - [ ] ZooKeeper
